### PR TITLE
matchtree: disable word search optimization for symbol search

### LINKED ISSUE #49892: search: case sensitive regex symbol search fails with andMatchTree inside query.Symbol
A regex symbol search with case sensitivity results in `error received from zoekt: found *zoekt.andMatchTree inside query.Symbol
`. EG [`type:symbol \bindexedData\b`](https://sourcegraph.com/search?q=context:global+type:symbol+%5CbindexedData%5Cb&patternType=regexp&case=yes&sm=1&groupBy=repo). A [customer](https://sourcegraph.slack.com/archives/C01D50MSA7L/p1679559269909209?thread_ts=1679426213.269819&cid=C01D50MSA7L) reported this happening on 4.5.1 but not on 4.1.3.

/cc @sourcegraph/search-team

### PR DESCRIPTION
**Pull request by keegancsmith**
Before this change case sensitive symbol searches of the form \bLITERAL\b would result in the error message

```
found *zoekt.andMatchTree inside query.Symbol
```

The root cause is the symbol search code is pretty janky. It constructs a matchtree and then pulls out either the regex matchtree or the substr matchtree, then creates a specific symbol searcher for those. It feels like it could be more generic rather than a bunch of hard to follow copy pasta. For now this was a more straightforward fix than creating a word search for symbols.

Test Plan: expanded the unit tests to cover word search more often. Additionally run zoekt-webserver and checked that a search like the following works now:

```
sym:\bnewMatchTree\b case:yes
```

Fixes https://github.com/sourcegraph/sourcegraph/issues/49892

### CODE REVIEW

File: matchtree.go [hunk 1]
```diff
@@ -963,7 +969,11 @@ func (d *indexData) newMatchTree(q query.Q) (matchTree, error) {
 		}, nil
 
 	case *query.Symbol:
-		subMT, err := d.newMatchTree(s.Expr)
+		// Disable WordMatchTree since we don't support it in symbols yet.
+		optCopy := opt
```
Comment (stefanhengl) [inline matchtree.go:973]: Why do we need to copy here? Doesn't NewMatchTree already get a shallow copy of opt?

**Review by stefanhengl — APPROVED** [review]

File: matchtree.go (hunk 1 above)
Comment (keegancsmith) [reply matchtree.go:973]: feels safer to avoid mutating, but yes in this case we can avoid copying. Gonna keep it this way since it feels safer to future mutations.
